torchmdnet/models/tensornet.py

```diff
@@ -480,18 +510,39 @@ def forward(
             edge_index, edge_attr[..., 2, None, None], S, X.shape[0]
         )
         msg = Im + Am + Sm
+        
+        prefactor = 1 if self.addtional_methods is not None else torch.ones_like(msg, device=msg.device, dtype=msg.dtype)
+        if self.addtional_methods is not None and extra_args is not None:
+            for label, method_dict in self.addtional_methods.items():
+                # appending to this list all the methods will be working in this way    
+                if method_dict['name'] in ['tensornet_q']:
+                    tmp_ = method_dict['method'](extra_args[label][..., None, None, None])
+                    #TODO: how do we want to handle prefactor if multiple methods are used here?
+                    prefactor +=  tmp_
+    
         if self.equivariance_invariance_group == "O(3)":
             A = torch.matmul(msg, Y)
             B = torch.matmul(Y, msg)
-            I, A, S = decompose_tensor((1 + 0.1 * q[..., None, None, None]) * (A + B))
+            I, A, S = decompose_tensor(prefactor * (A + B))
         if self.equivariance_invariance_group == "SO(3)":
             B = torch.matmul(Y, msg)
-            I, A, S = decompose_tensor(2 * B)
+            I, A, S = decompose_tensor(prefactor * 2 * B)
         normp1 = (tensor_norm(I + A + S) + 1)[..., None, None]
         I, A, S = I / normp1, A / normp1, S / normp1
         I = self.linears_tensor[3](I.permute(0, 2, 3, 1)).permute(0, 3, 1, 2)
         A = self.linears_tensor[4](A.permute(0, 2, 3, 1)).permute(0, 3, 1, 2)
-        S = self.linears_tensor[5](S.permute(0, 2, 3, 1)).permute(0, 3, 1, 2)
+        S = self.linears_tensor[5](S.permute(0, 2, 3, 1)).permute(0, 3, 1, 2) # shape: (natoms, hidden_channels, 3, 3)
         dX = I + A + S
-        X = X + dX + (1 + 0.1 * q[..., None, None, None]) * torch.matrix_power(dX, 2)
+        X = X + dX + (prefactor) * torch.matrix_power(dX, 2)
         return X
+
+
+class TensornetQ(nn.Module):
+  def __init__(self, init_value, additional_label='total_charge', learnable=False):
+    super().__init__()
+    self.prmtr = nn.Parameter(torch.tensor(init_value), requires_grad=learnable)
+    self.allowed_labels = ['total_charge', 'partial_charges']
+    assert additional_label in self.allowed_labels, f"Label {additional_label} not allowed for this method"
```

Comment:
Is there anything in particular preventing this method to be used with any other label apart from these two?
I could see it restricted to "just scalars", but other than that the method seems general.

Reply:
My original idea was to reintroduce only the additional_labels available now with this PR, but in theory the input could be anything